Adding a few basic tests

paulharte opened the pull request:
Adding a few very basic tests, hopefully this might encourage others to follow suit.

Also changed the changed the instantiation of the authentication object so that it doesn't happen at import time. (this would make testing very difficult). I made it a singleton, so that multiple objects are not created.

cuducos commented:
Many many thanks for popping in, @paulharte! That's a great start to #6 <3

In general, the code looks great, but I'd do minor edits in style before merging it. How do we prefer to proceed? I can (a) merge it locally, do the minor edits, and then merge your PR from my local branch; or (b) I can add inline comments on what I would change and you add a couple of commits addressing these issues?

In a nutshell, in _option a_ your job is done, and in _option b_ we can share & discuss the Python style issues I would work on. What do you prefer?

paulharte commented:
Hey Eduardo,

I am easy, if you have something specific in mind, go ahead and make the
changes. I am happy with whatever you want to do.
Paul



On Mon 30 Dec 2019, 21:18 Eduardo Cuducos, <notifications@github.com> wrote:

> Many many thanks for popping in, @paulharte <https://github.com/paulharte>!
> That's a great start to #6
> <https://github.com/cuducos/twitter-cleanup/issues/6> <3
>
> In general, the code looks great, but I'd do minor edits in style before
> merging it. How do we prefer to proceed? I can (a) merge it locally, do the
> minor edits, and then merge your PR from my local branch; or (b) I can add
> inline comments on what I would change and you add a couple of commits
> addressing these issues?
>
> In a nutshell, in *option a* your job is done, and in *option b* we can
> share & discuss the Python style issues I would work on. What do you prefer?
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/cuducos/twitter-cleanup/pull/19?email_source=notifications&email_token=AK7ZLLHVJ462U2WITQY2DMLQ3JQUFA5CNFSM4KBOR6A2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEH3HFNI#issuecomment-569799349>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AK7ZLLEXAGOZBD326ACE5ZDQ3JQUFANCNFSM4KBOR6AQ>
> .
>

cuducos commented on .github/workflows/black_and_test.yml:
```diff
@@ -1,4 +1,4 @@
-name: Black
+name: Black and Test
```
I would add another _workflow_ for tests instead of running them together with style check.

The rationale for that is that it would make feedback more straight forward: collaborators can know whether they've broken something (tests fail) or if it's just a matter of style (style checks fail)  without even looking in the CI logs.

[This repository](https://github.com/cuducos/bot-followers) might be an example of two different workflows for test and style check.

cuducos commented on .github/workflows/black_and_test.yml:
```diff
@@ -18,7 +18,13 @@ jobs:
     - name: Install dependencies
       run: |
         python -m pip install --upgrade pip
+        python setup.py install
```
Following [`pytest`'s best practices for integration with `setuptools`](http://doc.pytest.org/en/3.0.2/goodpractices.html#integrating-with-setuptools-python-setup-py-test-pytest-runner), this might be `pip install .` instead.

cuducos commented on .github/workflows/black_and_test.yml:
```diff
@@ -18,7 +18,13 @@ jobs:
         python -m pip install --upgrade pip
+        python setup.py install
         pip install black
+        pip install pytest
```
Following [`pytest`'s best practices for integration with `setuptools`](http://doc.pytest.org/en/3.0.2/goodpractices.html#integrating-with-setuptools-python-setup-py-test-pytest-runner), this might be removed (`python setup.py test` would sort that out, I guess).

cuducos commented on .github/workflows/black_and_test.yml:
```diff
@@ -18,7 +18,13 @@ jobs:
+
+    - name: Test with pytest
+      run: |
+        pytest
```
Following [`pytest`'s best practices for integration with `setuptools`](http://doc.pytest.org/en/3.0.2/goodpractices.html#integrating-with-setuptools-python-setup-py-test-pytest-runner), I prefer this to be `python setup.py test` (let's keep the standards, yay, hahaha). Also there's no need for the pipe is there's only one command/line: 

```yaml
    - name: Tests
       run: python setup.py test
```

cuducos commented on tests/test_twitterCleanup.py:
```diff
@@ -0,0 +1,26 @@
+from unittest import TestCase, mock
+import datetime
+
+import twitter_cleanup
```
I think that, overall, doing `from twitter_cleanup import TwitterCleanup` would grant us better readability in this file.

cuducos commented on tests/test_twitterCleanup.py:
```diff
@@ -0,0 +1,26 @@
+from unittest import TestCase, mock
+import datetime
```
I think that, overall, doing `from datetime import datetime` and `from unittest,mock import Mock, patch` would grant us better readability in this file.

Also, may we sort import in alphabetical order?

```
from datetime import datetime
from unittest import TestCase
from unittest.mock import Mock, patch
```

cuducos commented on tests/test_twitterCleanup.py:
```diff
@@ -0,0 +1,26 @@
+            status=mock.Mock(created_at=datetime.datetime(2019, 12, 2, 0, 0, 0, 0))
+        )
+        mock_user.status.text = "Hello"
+        mock_user.screen_name = "MrUser"
```
Is there a special reason you used the _argument_ architecture for the `.status.crated_at`, but the _attribute_ archtecture_ for other values? I mean, can we have the same strategy in mocking here?

Either:

```python
user = Mock(
    status=Mock(
        created_at=datetime(2019, 12, 2, 0, 0, 0, 0),
        text="Hello"
    ),
    screen_name="MrUser"
)
```

Or:

```python
user = Mock()
user.status.created_at = datetime(2019, 12, 2, 0, 0, 0, 0)
user.status.text = "Hello"
user.screen_name = "MrUser"
```

Personally I find the last one more readable and in line with what you suggested in `test_user.py` ; )

cuducos commented on tests/test_twitterCleanup.py:
```diff
@@ -0,0 +1,26 @@
+from unittest import TestCase, mock
```
`test_twitterCleanup.py` does not follow [PEP-8's module naming style](https://www.python.org/dev/peps/pep-0008/#package-and-module-names). Can we rename it as `test_twitter_cleanup.py`?

cuducos commented on tests/test_user.py:
```diff
@@ -0,0 +1,19 @@
+from unittest import TestCase, mock
+import datetime
```
I think that, overall, doing `from datetime import datetime` and `from unittest.mock import Mock, patch` would grant us better readability in this file.

Also, may we sort import in alphabetical order?

```
from datetime import datetime
from unittest import TestCase
from unittest.mock import Mock, patch
```

cuducos commented on tests/test_user.py:
```diff
@@ -0,0 +1,19 @@
+        user.status.created_at = datetime.datetime(2019, 12, 1, 0, 0, 0, 0)
+        self.assertTrue(user.last_status_before())
+
+        self.assertFalse(user.last_status_before(hours=48))
```
I would recommend one test case for each assertion here. The test method names would clarify the intention of each assertion and the test suite would be clearer for human readers. But this is up to you, I'm just thinking out loud here imagining something like `test_last_status_before_fails_if_post_older_than_today` hahaha...

Also, if you're refactoring this test case, you might want to take a look in [FreezeGun](https://pypi.org/project/freezegun/) ; )

cuducos commented on twitter_cleanup/authentication.py:
```diff
@@ -7,6 +7,15 @@
+    __instance = None
+
+    @staticmethod
+    def getInstance():
```
`getInstance` does not follow [PEP-8's naming conventions for methods](https://www.python.org/dev/peps/pep-0008/#method-names-and-instance-variables). May we have `get_instance` instead?

paulharte commented:
Hey @cuducos , I think my latest commit should take care off all the mentioned items. Have a happy New Year!

cuducos commented:
Yay — happy new year 🎉
